From 1c558ad7267450178fd434695d02ddcd36998d20 Mon Sep 17 00:00:00 2001
Message-Id: <1c558ad7267450178fd434695d02ddcd36998d20.1436852670.git.chang-joon.lee@intel.com>
In-Reply-To: <54e65a9ecec2815cf9cac81c7e32c2c6f7271166.1436852670.git.chang-joon.lee@intel.com>
References: <54e65a9ecec2815cf9cac81c7e32c2c6f7271166.1436852670.git.chang-joon.lee@intel.com>
From: Shubhangi Shrivastava <shubhangi.shrivastava@intel.com>
Date: Tue, 2 Jun 2015 10:42:24 +0530
Subject: [PATCH 2/4] REVERTME [VPG]: drm/i915: Enabling LPE Audio for Display
 Port

This patch enables LPE audio for DP by setting zeroth bit of
audio chickenbit register to allow programming through MMIO
debug registers. LPE audio then requires resetting the first
bit of audio port enable debug register to unmute the amplifier.

The audio port enable debug register is selected based on the
port on which DP is connected.

REVERTME: LP audio is not available upstream and may not be for
some more time which is required for this.

Issue: GMINL-10757
Change-Id: I5a4633db279316a3d85150f03524239defc323ae
Signed-off-by: Shubhangi Shrivastava <shubhangi.shrivastava@intel.com>
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
Signed-off-by: Durgadoss R <durgadoss.r@intel.com>
---
 drivers/gpu/drm/i915/hdmi_audio_if.c | 24 ++++++++++++++++++++++++
 drivers/gpu/drm/i915/i915_reg.h      | 13 +++++++++++++
 2 files changed, 37 insertions(+)

diff --git a/drivers/gpu/drm/i915/hdmi_audio_if.c b/drivers/gpu/drm/i915/hdmi_audio_if.c
index 5db1654..975d6f0 100644
--- a/drivers/gpu/drm/i915/hdmi_audio_if.c
+++ b/drivers/gpu/drm/i915/hdmi_audio_if.c
@@ -305,6 +305,8 @@ static int hdmi_audio_set_caps(enum had_caps_list set_element,
 	int ret = 0;
 	u32 hdmi_reg;
 	u32 int_masks = 0;
+	u32 chicken_bit;
+	u32 audio_enable;
 
 	DRM_DEBUG_DRIVER("\n");
 
@@ -314,6 +316,28 @@ static int hdmi_audio_set_caps(enum had_caps_list set_element,
 		if (hdmi_reg & PORT_ENABLE)
 			hdmi_reg |= SDVO_AUDIO_ENABLE;
 
+#ifdef CONFIG_SUPPORT_LPDMA_HDMI_AUDIO
+		if (IS_CHERRYVIEW(dev)) {
+			chicken_bit = I915_READ(VLV_AUD_CHICKEN_BIT_REG);
+			I915_WRITE(VLV_AUD_CHICKEN_BIT_REG,
+				chicken_bit | CHICKEN_BIT_DBG_ENABLE);
+
+			if (hdmi_priv->hdmi_reg == VLV_DP_B) {
+				audio_enable = I915_READ(VLV_AUD_PORT_EN_B_DBG);
+				I915_WRITE(VLV_AUD_PORT_EN_B_DBG,
+					audio_enable & ~AMP_UNMUTE);
+			} else if (hdmi_priv->hdmi_reg == VLV_DP_C) {
+				audio_enable = I915_READ(VLV_AUD_PORT_EN_C_DBG);
+				I915_WRITE(VLV_AUD_PORT_EN_C_DBG,
+					audio_enable & ~AMP_UNMUTE);
+			} else if (hdmi_priv->hdmi_reg == CHV_DP_D) {
+				audio_enable = I915_READ(VLV_AUD_PORT_EN_D_DBG);
+				I915_WRITE(VLV_AUD_PORT_EN_D_DBG,
+					audio_enable & ~AMP_UNMUTE);
+			}
+		}
+#endif
+
 		I915_WRITE(hdmi_priv->hdmi_reg, hdmi_reg);
 		I915_READ(hdmi_priv->hdmi_reg);
 		break;
diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 813e814..43c1be0 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -2998,6 +2998,19 @@ enum punit_power_well {
 #define   HDMI_MODE_SELECT_DVI			(0 << 9) /* HDMI only */
 #define   HDMI_COLOR_RANGE_16_235		(1 << 8) /* HDMI only */
 #define   SDVO_AUDIO_ENABLE			(1 << 6)
+
+/* Enabling LPE Audio */
+#define CHICKEN_BIT_DBG_ENABLE		(1 << 0)
+#define AMP_UNMUTE					(1 << 1)
+#define AUD_CHICKEN_BIT_REG			0x62F38
+#define AUD_PORT_EN_B_DBG			0x62F20
+#define AUD_PORT_EN_C_DBG			0x62F28
+#define AUD_PORT_EN_D_DBG			0x62F2C
+#define VLV_AUD_CHICKEN_BIT_REG		(VLV_DISPLAY_BASE + AUD_CHICKEN_BIT_REG)
+#define VLV_AUD_PORT_EN_B_DBG		(VLV_DISPLAY_BASE + AUD_PORT_EN_B_DBG)
+#define VLV_AUD_PORT_EN_C_DBG		(VLV_DISPLAY_BASE + AUD_PORT_EN_C_DBG)
+#define VLV_AUD_PORT_EN_D_DBG		(VLV_DISPLAY_BASE + AUD_PORT_EN_D_DBG)
+
 /* VSYNC/HSYNC bits new with 965, default is to be set */
 #define   SDVO_VSYNC_ACTIVE_HIGH		(1 << 4)
 #define   SDVO_HSYNC_ACTIVE_HIGH		(1 << 3)
-- 
1.9.1

